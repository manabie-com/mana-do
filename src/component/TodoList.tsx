import TodoItem from './TodoItem'
import React from 'react'
import { Todo } from '../models/todo'

function ToDoList({
  listTodos,
  onRemoveItem,
  filter,
  onUpdateData,
  onError,
}: {
  listTodos: any
  onRemoveItem: any
  filter: string
  onUpdateData: any
  onError: (mess: string) => void
}) {
  return (
    <div data-testid="list-todos">
      {listTodos?.length > 0 ? (
        listTodos.map((todo: Todo) => {
          return (
            <TodoItem
              onRemoveItem={onRemoveItem}
              key={todo.id + todo.status}
              data={todo}
              onUpdateData={onUpdateData}
              onError={onError}
            />
          )
        })
      ) : (
        <div>{filter ? `No data for filter ${filter}` : 'No data'}</div>
      )}
    </div>
  )
}

export default ToDoList

ToDoList.defaultProps = {
  listTodos: [],
  onRemoveItem: () => {},
  onUpdateData: () => {},
  filter: '',
  onError: () => {},
}
