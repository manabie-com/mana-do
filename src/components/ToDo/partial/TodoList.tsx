import React, { useState, useEffect } from 'react';

import { TodoStatus, EditTodo, Todo } from '../../../models/todo';
import { getRandomInt } from '../../../helpers';

import {
    updateTodo,
    updateTodoStatus,
    deleteTodo
} from '../../../store/actions';

import { MOTIV_MSG1, MOTIV_MSG2, MOTIV_MSG3, COMPLETED_TASK_MSG, NOT_YET_DONE_MSG } from '../../../constants';

const ALL: string = "ALL";
type EnhanceTodoStatus = TodoStatus | typeof ALL;

interface TodoListTypes {
    todos: Array<Todo>,
    showing: EnhanceTodoStatus,
    setShowing: any,
    dispatch: any
};

const motivationalMsg: string[] = [
    MOTIV_MSG1,
    MOTIV_MSG2,
    MOTIV_MSG3,
];

const TodoList = ({ todos, showing, setShowing, dispatch }: TodoListTypes) => {
    const [motivationMessage, setMotivationalMessage] = useState<string>(motivationalMsg[0]);
    const [editTodo, setEditTodo] = useState<EditTodo>({ editMode: false, id: null, content: null });

    useEffect(() => {
        if (showing === TodoStatus.COMPLETED) {
            setMotivationalMessage(motivationalMsg[getRandomInt(3)]);
        }
    }, [showing]);

    const onDeleteTodoById = (todiId: any) => dispatch(deleteTodo(todiId));

    const onChangeEditTodo = (id: any, content: any) => setEditTodo(prev => ({ ...prev, id, content }));

    const handleEditTodo = (id: any, content: any) => setEditTodo({ editMode: true, id, content });

    const onUpdateTodo = (id: any, content: any) => {
        let data = { editMode: false, id, content };
        setEditTodo(prev => ({ ...prev, ...data }));
        dispatch(updateTodo(data));
    };

    const onUpdateTodoStatus = (e: React.ChangeEvent<HTMLInputElement>, todoId: any) => dispatch(updateTodoStatus(todoId, e.target.checked));

    const filterTodosList = (status: string) => {
        if (showing === ALL) {
            return true
        } else if (showing === status) {
            return true;
        } else {
            return false;
        }
    };

    const checkTodosByStatus = () => {
        const result = todos.find(item => item.status === showing);

        if (!result && showing !== ALL) {
            if (todos.length > 0) {
                if (showing === TodoStatus.COMPLETED) {
                    return <>
                        <small>{NOT_YET_DONE_MSG}</small>
                        <p>{motivationMessage}</p>
                    </>;
                } else {
                    return COMPLETED_TASK_MSG;
                }
            }
        }
    };

    return (
        <div className="ToDo__list">
            {   
                todos.map((todo, index) =>
                    filterTodosList(todo.status) &&
                    <div
                        key={index}
                        className="ToDo__item"
                        data-testid="todo__item"
                    >
                        <input
                            type="checkbox"
                            checked={todo.status === TodoStatus.COMPLETED}
                            onChange={e => onUpdateTodoStatus(e, todo.id)}
                        />
                        <span className={`todo__${todo.status}`} onDoubleClick={() => handleEditTodo(todo.id, todo.content)}>
                            {
                                (editTodo.editMode && editTodo.id === todo.id)
                                ? <input
                                    value={editTodo.content}
                                    onChange={e => onChangeEditTodo(todo.id, e.target.value)}
                                    onBlur={() => onUpdateTodo(todo.id, editTodo.content)}
                                    style={{ width: '95%' }}
                                />
                                : todo.content
                            }
                        </span>
                        <button
                            className="Todo__delete"
                            onClick={() => onDeleteTodoById(todo.id)}
                        >
                            X
                        </button>
                    </div>
                )
            }
            {checkTodosByStatus()}
            {
                todos.length === 0
                && <div
                        className='toDo__message'
                        data-testid='no__todo__message'
                    >
                        <h4>MANABIE TO DO</h4>
                        <p>Free up your mental space, gives you focus, from work to play.</p>
                    </div>
            }
        </div>
    )
};

export default TodoList;