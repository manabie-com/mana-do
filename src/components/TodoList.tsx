import React, {Dispatch, SetStateAction} from "react";
import {AppActions,} from "store/actions";
import {Todo, TodoStatus} from "models/todo";
import TodoItem from "./TodoItem";

type EnhanceTodoStatus = TodoStatus | 'ALL';

interface ToolBarProps {
    dispatch: Dispatch<AppActions>,
    todos: Array<Todo>,
    showing: EnhanceTodoStatus,
    setShowing: Dispatch<SetStateAction<EnhanceTodoStatus>>
}

const TodoList: React.FC<ToolBarProps> = ({dispatch, todos, showing, setShowing}) => {
    const todosFilter = todos.filter((todo) => {
        switch (showing) {
            case TodoStatus.ACTIVE:
                return todo.status === TodoStatus.ACTIVE;
            case TodoStatus.COMPLETED:
                return todo.status === TodoStatus.COMPLETED;
            default:
                return true;
        }
    });

    return (
        <div className="Todo__list">
            {todosFilter.map((todo, index) => {
                const todoItemProps = {dispatch, todo}
                return <TodoItem key={todo.id} {...todoItemProps} />;
            })}
            {todos.length > 0 && !(todosFilter.length > 0) &&
            <div className="__no-todos">{"no " + (showing === TodoStatus.ACTIVE ? 'active' : 'completed') + " todos..."}</div>}
        </div>
    );
};

export default TodoList;
