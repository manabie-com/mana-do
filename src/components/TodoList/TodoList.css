.app-table {
    border-collapse: collapse;
    width: 100%;
    table-layout: fixed;

}   

.app-table thead tr {
    background-color: #39a1f4;
    color: #fff;

}

.app-table th,
.app-table td {
    padding: 5px 24px;
}

.app-table td {
    white-space:pre-wrap; 
    word-wrap:break-word
}

.app-table tbody tr {
    border-bottom: 1px solid #dddddd;
}

.app-table .th-15-percent {
    width: 15%;
}

.app-table .th-seq-number {
    width: 1%;
}

.app-table .align-center {
    text-align: center;
}

.app-table .align-left {
    text-align: left;
}

.app-table .align-right {
    text-align: right;
}