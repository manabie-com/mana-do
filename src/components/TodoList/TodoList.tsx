import React from "react";

import { Todo } from "types";
import TodoItem from "components/TodoItem";

type PropTypes = {
  todos: Todo[];
};

export default function TodoList({ todos = [] }: PropTypes): JSX.Element {
  return (
    <div data-testid="todo-list">
      {todos.map(({ id, completed, content }) => (
        <TodoItem key={id} id={id} content={content} isCompleted={completed} />
      ))}
    </div>
  );
}
