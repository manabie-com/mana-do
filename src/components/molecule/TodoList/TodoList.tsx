import { Todo, TodoStatus } from "models/todo";
import React from "react";
import ToDoItem from "components/atom/TodoItem";

export type EnhanceTodoStatus = TodoStatus | "ALL";

type ToDoListProps = {
  showTodos: Todo[];
  showing: EnhanceTodoStatus;
};

export const emptyTodoMessage = (showing: EnhanceTodoStatus) =>
  `There are not${
    showing === TodoStatus.ACTIVE
      ? " active"
      : showing === TodoStatus.COMPLETED
      ? " completed"
      : ""
  } items`;

const ToDoList: React.FC<ToDoListProps> = ({ showTodos, showing }) => {
  return (
    <div className="ToDo__list">
      {showTodos.length === 0 && (
        <span className="Todo__empty__message" data-testid="empty-message">
          {emptyTodoMessage(showing)}
        </span>
      )}
      {showTodos.map((todo) => {
        return <ToDoItem todo={todo} key={todo.id} />;
      })}
    </div>
  );
};

export default ToDoList;
