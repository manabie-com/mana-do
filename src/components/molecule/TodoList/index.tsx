import ToDoList from "./TodoList";

export default ToDoList;
