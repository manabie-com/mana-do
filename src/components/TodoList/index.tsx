import React, { useContext } from "react";
import { AppContext } from "../../store/context";
import TodoItem from './TodoItem';
import './TodoList.css';
interface ITodoListProps {
    filter:string,
    [x:string]:string | any
}

const TodoList = ({filter}:ITodoListProps) => {
    const { state } = useContext(AppContext);
    const { todos } = state; 
    const filteredTodo = todos.filter(item => filter!=='ALL'?item.status === filter:item);
    return (<div className="table-container"><table className="app-table">
            <thead>
                <tr>
                    <th className="th-seq-number align-center">No</th>
                    <th className="th-seq-number align-center">Done?</th>
                    <th className="th-300px align-left">Content</th>
                    <th className="th-100px align-center">Status</th>
                    <th className="th-100px align-center">Created</th>
                    <th className="th-160px align-center">Action</th>
                </tr>
            </thead>
            <tbody>
                    { filteredTodo.map((item,index)=>{
                        return <TodoItem itemNo={index+1} key={item.id} todo={item}  />
                    }
                    ) }
            </tbody>
        </table></div>)
}

export default TodoList;


