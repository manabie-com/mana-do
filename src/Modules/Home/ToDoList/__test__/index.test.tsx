import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import TodoList from "../index";

describe('<TodoList/> render', () => {
        it('should match snapshot', function () {
                const loadingIndicator = render(<TodoList/>);
                expect(loadingIndicator.container.firstChild).toMatchSnapshot();
        });

        it('shows success message after confirm button is clicked', () => {
        })
})