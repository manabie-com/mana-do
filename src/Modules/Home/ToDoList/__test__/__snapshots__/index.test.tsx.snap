// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList/> render should match snapshot 1`] = `
<div
  class="bar bar__left"
>
  <div
    class="WrapperItemSiderBar"
  >
    <div
      class="ItemSiderBar--acitve ItemSiderBar"
    >
      <i
        class="manabie-endless"
      />
      <span>
        All
      </span>
    </div>
  </div>
  <div
    class="WrapperItemSiderBar"
  >
    <div
      class="false ItemSiderBar"
    >
      <i
        class="manabie-active"
      />
      <span>
        Active
      </span>
    </div>
  </div>
  <div
    class="WrapperItemSiderBar"
  >
    <div
      class="false ItemSiderBar"
    >
      <i
        class="manabie-phong"
      />
      <span>
        Completed
      </span>
    </div>
  </div>
  <div
    class="WrapperItemSiderBar"
  >
    <div
      class="false ItemSiderBar"
    >
      <i
        class="manabie-star_outline"
      />
      <span>
        Important
      </span>
    </div>
  </div>
  <div
    class="WrapperItemSiderBar"
  >
    <div
      class="false ItemSiderBar"
    >
      <i
        class="manabie-remove"
      />
      <span>
        Clear all todos
      </span>
    </div>
  </div>
</div>
`;
